client/sections-middleware.js

```diff
@@ -18,66 +18,89 @@ import { receiveSections, load } from './sections-helper';
 import sections from './sections';
 receiveSections( sections );
 
-const _loadedSections = {};
-
-function activateSection( sectionDefinition, context, next ) {
-	const dispatch = context.store.dispatch;
-
+function activateSection( sectionDefinition, context ) {
 	controller.setSection( sectionDefinition )( context );
-	dispatch( { type: 'SECTION_SET', isLoading: false } );
-	dispatch( activateNextLayoutFocus() );
-	next();
+	context.store.dispatch( activateNextLayoutFocus() );
 }
 
-function createPageDefinition( path, sectionDefinition ) {
-	const pathRegex = pathToRegExp( path );
+async function loadSection( context, sectionDefinition ) {
+	context.store.dispatch( { type: 'SECTION_SET', isLoading: true } );
 
-	page( pathRegex, function( context, next ) {
-		const envId = sectionDefinition.envId;
-		const dispatch = context.store.dispatch;
+	// If the section chunk is not loaded within 400ms, report it to analytics
+	const loadReportTimeout = setTimeout( () => {
+		context.store.dispatch( bumpStat( 'calypso_chunk_waiting', sectionDefinition.name ) );
+	}, 400 );
 
-		if ( envId && envId.indexOf( config( 'env_id' ) ) === -1 ) {
-			return next();
-		}
+	try {
+		// load the section module, i.e., its webpack chunk
+		const requiredModule = await load( sectionDefinition.name, sectionDefinition.module );
+		// call the module initialization function (possibly async, registers page.js handlers etc.)
+		await requiredModule.default( controller.clientRouter );
+	} finally {
+		context.store.dispatch( { type: 'SECTION_SET', isLoading: false } );
 
-		if ( _loadedSections[ sectionDefinition.module ] === true ) {
-			return activateSection( sectionDefinition, context, next );
-		}
+		// If the load was faster than the timeout, this will cancel the analytics reporting
+		clearTimeout( loadReportTimeout );
+	}
+}
 
-		dispatch( { type: 'SECTION_SET', isLoading: true } );
+/**
+ * Cache of already loaded or loading section modules. Every section module is in one of
+ * three states regarding the cache:
+ * - no record in the cache: not loaded or not currently loading
+ * - record value is `true`: already loaded and initialized
+ * - record value is a `Promise`: is currently loading, the promise will fulfill when done.
+ *   Don't start a second load with `loadSection` but rather wait for the existing promise.
```

Comment:
thought: what if we don't use `true` to indicate a section is already loaded but reuse the (already fulfilled) promise? does that make sense at all?

Reply:
If we use `true`, it means that we can call `next` and continue to the next page.js handler synchronously. With `resolvedPromise.then`, we always wait for the next event loop tick. Not a big deal at all, but still nice to have.

My code was influenced by how webpack implements the `requireEnsure` function that loads chunks that are not loaded yet, especially the `installedChunks` cache. The source is [here](https://github.com/webpack/webpack/blob/614a6fc0109a702ee64ca0b0982a0d638b1b6ca1/lib/web/JsonpMainTemplatePlugin.js#L259-L294) in a hard-to-read form. It's easier to read the actual generated code in the webpack runtime (we call it `manifest` in Calypso).

Reply:
thanks @jsnajdr makes sense 👍